src/wallet/wallet.cpp

```diff
@@ -470,6 +524,81 @@ void CWallet::AddToSpends(const uint256& wtxid)
         AddToSpends(txin.prevout, wtxid);
+    BOOST_FOREACH(const JSDescription& jsdesc, thisTx.vjoinsplit)
+        BOOST_FOREACH(const uint256& nullifier, jsdesc.nullifiers)
+            AddToSpends(nullifier, wtxid);
+}
+
+void CWallet::IncrementNoteWitnesses(const CBlockIndex* pindex, const CBlock* pblockIn)
+{
+    {
+        LOCK(cs_wallet);
+        for (NoteWitnesses::value_type& item : mapNoteWitnesses) {
+            // Copy the witness for the previous block
+            item.second.push_front(item.second.front());
+            // Cache enough witnesses that we can expect
+            // to never reorg beyond our cache.
+            if (item.second.size() > 50)
+                item.second.pop_back();
+        }
+
+        const CBlock* pblock {pblockIn};
+        CBlock block;
+        if (!pblock) {
+            ReadBlockFromDisk(block, pindex);
+            pblock = &block;
+        }
+
+        ZCIncrementalMerkleTree tree;
+        bool treeInitialised = false;
+        for (const CTransaction& tx : pblock->vtx) {
+            if (!treeInitialised && tx.vjoinsplit.size() > 0) {
+                LOCK(cs_main);
+                if (!pcoinsTip->GetAnchorAt(tx.vjoinsplit[0].anchor, tree)) {
+                    // This should not happen, because IncrementNoteWitnesses()
+                    // is only called when the chain tip updates, and the
+                    // anchors for the JoinSplits in that block should still be
+                    // cached.
+                    // TODO: Calculate the anchor from scratch?
+                    throw std::runtime_error("CWallet::IncrementNoteWitnesses(): anchor not cached");
+                }
+                treeInitialised = true;
+            }
+
+            auto hash = tx.GetTxid();
+            for (uint8_t i = 0; i < tx.vjoinsplit.size(); i++) {
+                const JSDescription& jsdesc = tx.vjoinsplit[i];
+                for (uint8_t j = 0; j < jsdesc.commitments.size(); j++) {
+                    const uint256& note_commitment = jsdesc.commitments[j];
+                    tree.append(note_commitment);
+
+                    // Increment existing witnesses
+                    for (NoteWitnesses::value_type& item : mapNoteWitnesses) {
+                        item.second.front().append(note_commitment);
+                    }
+
+                    // If this is our note, witness it
+                    if (mapWallet.count(hash)) {
+                        JSOutPoint jsoutpt {hash, i, j};
+                        if (mapWallet[hash].mapNoteData.count(jsoutpt)) {
+                            mapNoteWitnesses[jsoutpt].push_front(tree.witness());
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+void CWallet::DecrementNoteWitnesses()
+{
+    {
+        LOCK(cs_wallet);
+        for (NoteWitnesses::value_type& item : mapNoteWitnesses) {
+            if (item.second.size() > 0)
+                item.second.pop_front();
```

Comment:
There are two cases where this could be empty:

- The block in which the note was spent has been removed.
- The note is older than the cache window, and a reorg has occurred that has wiped out the entire cache.

In the former case, we *want* this list to be empty. In the latter case (which should occur very rarely depending on the cache window size), we'd need to regenerate the cache from scratch; I'll open a ticket for implementing this.